### PR TITLE
Moves out node pool manager behind an extension using runner pool abstraction (Part 2)

### PR DESCRIPTION
**Pull request by gviedma**

### CODE REVIEW

File: api/agent/lb_agent.go [hunk 1]
```diff
@@ -129,22 +92,23 @@ const (
...
 	// amount of time to wait to place a request on a runner
-	placementTimeout = 15 * time.Second
+	placementTimeout          = 15 * time.Second
+	runnerPoolShutdownTimeout = 5 * time.Second
 )
 
 type lbAgent struct {
 	delegatedAgent Agent
-	np             NodePool
+	rp             models.RunnerPool
```
Comment (zootalures) [inline api/agent/lb_agent.go]: Sorry I should have said this on the previous PR - is `models.` the right package for the RunnerPool? - models would seem to be the place for things related to persistence and exchange of data, RunnerPool would seem to either sit on its own or somewhere else.

File: api/agent/static_pool.go [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+func (np *staticRunnerPool) Runners(call models.RunnerCall) ([]models.Runner, error) {
+	np.rMtx.RLock()
+	defer np.rMtx.RUnlock()
+
+	r := make([]models.Runner, len(np.runners))
+	copy(r, np.runners)
+	return r, nil
+}
+
+func (np *staticRunnerPool) AddRunner(address string) error {
```
Comment (ostrain) [inline api/agent/static_pool.go]: Minor nit, but what about `rp` rather than `np` for the receiver? (for this method as well as the others in this file)

**Comment by ostrain** [comment]
Oh, one other question: do we need new extension point(s)? I’m not really familiar with the mechanics of extensions, but I didn’t see anything along those lines.

File: api/agent/lb_agent.go (hunk 1 above)
Comment (gviedma) [reply api/agent/lb_agent.go]: I moved all interfaces to api/runnerpool.

File: docs/operating/runner_pools.md [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+```bash
+mkdir /tmp/runnerdata
+FN_NODE_TYPE=pure-runner FN_PORT=8082 FN_GRPC_PORT=9190 ./fnserver
+```
+
+### LB
+
+```bash
+mkdir /tmp/lbdata
+FN_NODE_TYPE=lb FN_PORT=8081 FN_RUNNER_API_URL=http://localhost:8080 FN_RUNNER_ADDRESSES=localhost:9190 FN_LOG_LEVEL=DEBUG ./fnserver
```
[the remark this answers is not in the document]
Comment (gviedma) [reply docs/operating/runner_pools.md]: I've added a second runner to the example.

File: api/agent/lb_agent_test.go [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+
+func (r *mockRunner) Close(ctx context.Context) error {
+	go func() {
+		r.wg.Wait()
+	}()
+	return nil
+}
+
+func (r *mockRunner) Address() string {
+	return ""
```
[the remark this answers is not in the document]
Comment (gviedma) [reply api/agent/lb_agent_test.go]: Done

File: api/agent/lb_agent_test.go [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+				failures <- fmt.Errorf("Timed out call %d", i)
+			}
+		}(i)
+	}
+
+	wg.Wait()
+	close(failures)
+
+	err := <-failures
+	if err != nil {
```
[the remark this answers is not in the document]
Comment (gviedma) [reply api/agent/lb_agent_test.go]: Now we explicitly verify that the spillover occurred by tracking the number of processed calls per runner.

File: api/agent/runner_client.go [hunk 6]
```diff
@@ -0,0 +1,249 @@
...
+func (r *gRPCRunner) Address() string {
+	return r.address
+}
+
+func (r *gRPCRunner) TryExec(ctx context.Context, call models.RunnerCall) (bool, error) {
+	logrus.WithField("runner_addr", r.address).Debug("Attempting to place call")
+	r.wg.Add(1)
+	defer r.wg.Done()
+
+	// Get app and route information
```
[the remark this answers is not in the document]
Comment (gviedma) [reply api/agent/runner_client.go]: Cleaned up the comment.

File: api/agent/static_pool.go (hunk 2 above)
Comment (gviedma) [reply api/agent/static_pool.go]: Done

File: api/agent/runner_client.go [hunk 7]
```diff
@@ -0,0 +1,249 @@
...
+			logrus.WithError(err).Error("Failed to send data frame")
+			return err
+		}
+	}
+	return nil
+}
+
+func receiveFromRunner(protocolClient pb.RunnerProtocol_EngageClient, c models.RunnerCall, done chan error) {
+	w, ok := c.ResponseWriter().(http.ResponseWriter)
+	if !ok {
```
[the remark this answers is not in the document]
Comment (gviedma) [reply api/agent/runner_client.go]: Good point. I've pushed it up to the RunnerCall interface to always return an _http.ResponseWriter_. That should always be the case when running in a load-balanced setup.

**Pull request merged**